### ISSUE TITLE
Scheduler not dispatching to workers

### ISSUE THREAD

**Issue by tk3369** [JuliaParallel/Dagger.jl#84]
This may be the same issue as #71 .

I have 3 workers and  I'm using Julia 0.6.2.

This simple graph has 3 leaf thunks (`x2`) and 1 master thunk (`x3`).  It works properly.
```
julia> x2 = [delayed(_ -> (sleep(2); println(i)))(nothing) for i in 1:3]
3-element Array{Dagger.Thunk,1}:
 *29*
 *30*
 *31*

julia> x3 = delayed((x...) -> println("done: ", x))(x2...)
*32*

julia> @time compute(x3)
	From worker 3:	2
	From worker 4:	3
	From worker 2:	1
	From worker 2:	done: (nothing, nothing, nothing)
  2.320594 seconds (21.23 k allocations: 1.159 MiB)
Dagger.Chunk{Void,MemPool.DRef}(Void, Dagger.UnitDomain(), MemPool.DRef(2, 39, 0x0000000000000000), false)
```

This case is broken.  The only difference is that the new leaf thunk (`x1`) was added for which `x2` thunks depend on.

```
julia> x1 = delayed(x -> (sleep(2); println("root")))(1)
*33*

julia> x2 = [delayed(_ -> (sleep(2); println(i)))(x1) for i in 1:3]
3-element Array{Dagger.Thunk,1}:
 *34*
 *35*
 *36*

julia> x3 = delayed((x...) -> println("done: ", x))(x2...)
*37*

julia> @time compute(x3)
	From worker 2:	root
	From worker 2:	1
	From worker 2:	2
	From worker 2:	3
	From worker 2:	done: (nothing, nothing, nothing)
  8.153534 seconds (31.29 k allocations: 1.678 MiB)
Dagger.Chunk{Void,MemPool.DRef}(Void, Dagger.UnitDomain(), MemPool.DRef(2, 44, 0x0000000000000000), false)
```

**Comment by jpsamaroo** [comment]
Based on adding some debug statements to the scheduler, I think what's happening here is the following:

- `x1` runs to completion.
- `x2[1]` begins executing and is placed in `state.running`, and `x2[2]` and `x2[3]` are placed in `state.ready`. This is the first misstep by the scheduler, as it's clear that `x2[2]` and `x2[3]` can be run. Probably the same issue as #53 
- At the top of the scheduler loop, https://github.com/JuliaParallel/Dagger.jl/blob/4dd1fe690660c809b9065d116f6f2c7abac4235f/src/scheduler.jl#L45 because a thunk is in `state.running`, the scheduler skips dispatching more thunks and waits for `x2[1]` to complete. This is the second misstep by the scheduler, since again, `x2[2]` and `x2[3]` can be run at any time.
- The cycle repeats until we get to `x3`.

So, the root of this issue is likely that the node which contains the data (result of `x1`) that's needed for the other 2 pending-but-ready tasks is tied up with executing a thunk and is unable to either accept another next thunk from the scheduler, or serialize results. Additionally, the scheduler is much too conservative about eagerly scheduling more work while other thunks are running; although maybe this is a defence mechanism to prevent unintended behavior in some weird corner cases, like where the input data is mutated (don't do this, please make copies of your inputs if you plan to mutate them!).

The solution in my mind is to first make the scheduler less conservative at https://github.com/JuliaParallel/Dagger.jl/blob/4dd1fe690660c809b9065d116f6f2c7abac4235f/src/scheduler.jl#L45 and see if this breaks anything; we'll probably not want to schedule on the same worker that's actively running, so we can check which workers are busy with thunks in `state.running` and exclude them from receiving new thunks.

However, the above won't do anything useful if the data-owning worker is too busy to send its data to the worker that needs it; I propose that we have the scheduler do some data-dependency checks on a thunk `A` before dispatching it, and if we find that more than one thunk depends on the result of thunk `A`, we have the worker that completes `A` sit in "time-out" until other workers can pull the result from it. This is a slightly ugly solution, but once I/O is thread-safe in Julia, we can consider setting aside an I/O thread to service requests for cached data from other workers. The other alternative is something akin to torrenting, where we could ensure that every thunk's result is mirrored to a few other workers, and a worker cannot accept a new thunk until the result is either no longer needed, or another worker with that cached data "agrees" to sit and serve it to other running thunks that need it.

**Comment by Losses** [comment]
Hmmm... any progress about this?

**Comment by jpsamaroo** [comment]
@shashi had recently prepared some multithreading changes to Dagger and MemPool might support the solution I detailed above. In short, what's needed is the ability for a worker to be able to serialize data while it's working on a (potentially unrelated) thunk, which will require the new PARTR MT model. I will try to modify the scheduler to make it more liberal with scheduling and see if the MT changes make that work well.

**Comment by Losses** [comment]
Wow, that's cool, I'll use it on my personal project :D

**Comment by jpsamaroo** [comment]
I've confirmed that with https://github.com/JuliaComputing/MemPool.jl/pull/35, Shashi's `s/threads` branch, and a small one-line tweak to the scheduler, we start running more than one thread on each worker in parallel using threads. However, I don't know yet how well this scales w.r.t running large computations alongside serializing large datasets (common in HPC, where they often do matrix ops on huge matrices). Once JuliaComputing/MemPool.jl#35 is merged and Shashi merges his PR, I'll dive in deeper and try to make things as efficient as I can.

**Comment by jpsamaroo** [comment]
Fixed by #177
